### ISSUE TITLE
gnutls patch sha256 seemingly incorrect

### ISSUE THREAD

**Issue by JakeWharton** [Homebrew/homebrew-core#43156]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] are reporting a bug others will be able to reproduce and not asking a question or requesting software. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh. To get software added or changed in Homebrew please file a [Pull Request](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)
- [x] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, Homebrew/homebrew-core formula (not cask) on macOS? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new/choose. If it's a Linux problem please file this issue at https://github.com/Homebrew/linuxbrew-core/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?

```
$ brew gist-logs gnutls
Error: No logs.

$ brew gist-logs gnupg
Error: Not Found
```

- [x] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

```
$ brew config
HOMEBREW_VERSION: 2.1.10
ORIGIN: https://github.com/Homebrew/brew
HEAD: 925ec737fba1db97d0170c56de982900711cacd3
Last commit: 3 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 3d417df47f02fba0c118678916aeef0656a0360e
Core tap last commit: 6 minutes ago
HOMEBREW_PREFIX: /Users/jakew/dev/other/homebrew
HOMEBREW_CELLAR: /Users/jakew/dev/other/homebrew/Cellar
HOMEBREW_REPOSITORY: /Users/jakew/dev/other/homebrew
CPU: octa-core 64-bit kabylake
Homebrew Ruby: 2.3.7 => /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby
Clang: 10.0 build 1001
Git: 2.22.0 => /Users/jakew/dev/other/homebrew/bin/git
Curl: 7.54.0 => /usr/bin/curl
Java: 14, 13, 12.0.1, 11.0.3, 10, 9.0.1, 1.8.0_212, 1.7.0_80, 1.6.0_65-b14-468
macOS: 10.14.5-x86_64
CLT: 10.3.0.0.1.1562985497
Xcode: 10.2.1
CLT headers: 10.3.0.0.1.1562985497

$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Your Homebrew's prefix is not /usr/local.
Some of Homebrew's bottles (binary packages) can only be used with the default
prefix (/usr/local).
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Discourse, Twitter or IRC. You are responsible for resolving any issues you
experience, as you are running this unsupported configuration.


Warning: Homebrew's sbin was not found in your PATH but you have installed
formulae that put executables in /Users/jakew/dev/other/homebrew/sbin.
Consider setting the PATH for example like so:
  echo 'export PATH="/Users/jakew/dev/other/homebrew/sbin:$PATH"' >> ~/.bash_profile
```

To help us debug your issue please explain:

- What you were trying to do (and why)

Regular `brew upgrade` with `gnupg` already installed at version 2.2.16_1

- What happened (include command output)

```
$ brew upgrade
==> Upgrading 5 outdated packages:
gnupg 2.2.16_1 -> 2.2.17, repo 1.13.3 -> 1.13.4, git 2.22.0_1 -> 2.22.1, buildifier 0.26.0 -> 0.28.0, gnutls 3.6.8 -> 3.6.9
==> Upgrading gnupg
==> Installing dependencies for gnupg: gnutls
==> Installing gnupg dependency: gnutls
==> Downloading https://www.gnupg.org/ftp/gcrypt/gnutls/v3.6/gnutls-3.6.9.tar.xz
######################################################################## 100.0%
==> Downloading https://gitlab.com/gnutls/gnutls/commit/ef80617d1e17e0878a909baad62a75ba265c0e00.patch
######################################################################## 100.0%
Error: An exception occurred within a child process:
  ChecksumMismatchError: SHA256 mismatch
Expected: 3b1634fa348c0f0064e43f3fb673e30a5e46f2b51cd6cd0d4a0cbf326e71c90e
  Actual: 34480b13120e2167060795060b6738c26622c341498cf6303b17b143642b14cd
 Archive: /Users/jakew/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch
To retry an incomplete download, remove the file above.
```

Deleted the patch file and re-ran, same result. Deleted the patch file and the gnutls 3.6.9 tar.xz file and re-ran, same result.

- What you expected to happen

`gnutls` 3.6.9 installs successfully allowing `gnupg` to install.

- Step-by-step reproduction instructions (by running `brew install` commands)

Given an installation of `gnupg` 2.2.16_1 and `gnutls` 3.6.8 (and any other required dependencies), attempt `brew upgrade`.

---

The formula for `gnutls` currently contains:

https://github.com/Homebrew/homebrew-core/blob/2f555dbfbcd65888c8deb00bb22b16926de7e698/Formula/gnutls.rb#L31-L34

When downloading that patch manually and piping through `sha256sum` I see the same result as the failed install which is different from the formula value:

```
$ curl https://gitlab.com/gnutls/gnutls/commit/ef80617d1e17e0878a909baad62a75ba265c0e00.patch | sha256sum
34480b13120e2167060795060b6738c26622c341498cf6303b17b143642b14cd
```

**Comment by SMillerDev** [comment]
Thanks for the detailed issue. The patch of a commit changing is very weird. Could you make a pull request changing the hash? We'll then continue the discussion there. (I'm not able to make a pull request for you right now)

**Comment by JakeWharton** [comment]
Sure!

**Comment by jkkm** [comment]
Amusingly, it appears to be caused by gitlab changing the backend git version used to generate these patches:

```
[jkkm@cygnus ~]$ curl -O https://gitlab.com/gnutls/gnutls/commit/ef80617d1e17e0878a909baad62a75ba265c0e00.patch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2378    0  2378    0     0   3866      0 --:--:-- --:--:-- --:--:--  3866
[jkkm@cygnus ~]$ sha256sum ef80617d1e17e0878a909baad62a75ba265c0e00.patch 
34480b13120e2167060795060b6738c26622c341498cf6303b17b143642b14cd  ef80617d1e17e0878a909baad62a75ba265c0e00.patch
[jkkm@cygnus ~]$ sed -e 's!2.22.0!2.21.0!' -i ef80617d1e17e0878a909baad62a75ba265c0e00.patch 
[jkkm@cygnus ~]$ sha256sum ef80617d1e17e0878a909baad62a75ba265c0e00.patch 
3b1634fa348c0f0064e43f3fb673e30a5e46f2b51cd6cd0d4a0cbf326e71c90e  ef80617d1e17e0878a909baad62a75ba265c0e00.patch
[jkkm@cygnus ~]$ 
```

**Comment by Varshul** [comment]
Thanks @jkkm. This worked like a charm.

Just remove the existing downloaded patch and copy this to the same file name and everybody goes home happy :)

**Comment by RgRaam** [comment]
Hello there, @JakeWharton @jkkm @Varshul . 
I am facing the same issue here. I am beginner and so can you explain it detailed as i am not sure on what to do to disclose this error. On Mac 10.11.6;

 
```
==> Downloading https://www.gnupg.org/ftp/gcrypt/gnutls/v3.6/gnutls-3.6.9.tar.xz
Already downloaded: /Users/apple/Library/Caches/Homebrew/downloads/10375e457c78424d079bc03dc50b19e2d8f153be4087edde62a035a2627d8c6f--gnutls-3.6.9.tar.xz
==> Downloading https://gitlab.com/gnutls/gnutls/commit/ef80617d1e17e0878a909baa
Already downloaded: /Users/apple/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch
Error: An exception occurred within a child process:
  ChecksumMismatchError: SHA256 mismatch
Expected: 3b1634fa348c0f0064e43f3fb673e30a5e46f2b51cd6cd0d4a0cbf326e71c90e
  Actual: 34480b13120e2167060795060b6738c26622c341498cf6303b17b143642b14cd
 Archive: /Users/apple/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch
To retry an incomplete download, remove the file above.

```

**Comment by hnry** [comment]
@RgRaam edit the patch file, should be `/Users/apple/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch` with nano, vim or whatever editor you'd like.

The only change you need to make is in the last line where it says "2.22.0" , change that to "2.21.0" (changing the 22 to 21)

Save and re-run your brew install

**Comment by RgRaam** [comment]
> @RgRaam edit the patch file, should be `/Users/apple/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch` with nano, vim or whatever editor you'd like.
> 
> The only change you need to make is in the last line where it says "2.22.0" , change that to "2.21.0" (changing the 22 to 21)
> 
> Save and re-run your brew install


Thank You for the Reply @hnry . But it din't Solve. But it came with another mismatch error.


`==> Downloading https://www.gnupg.org/ftp/gcrypt/gnutls/v3.6/gnutls-3.6.9.tar.xz
Already downloaded: /Users/apple/Library/Caches/Homebrew/downloads/10375e457c78424d079bc03dc50b19e2d8f153be4087edde62a035a2627d8c6f--gnutls-3.6.9.tar.xz
==> Downloading https://gitlab.com/gnutls/gnutls/commit/ef80617d1e17e0878a909baa
Already downloaded: /Users/apple/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch
Error: An exception occurred within a child process:
  ChecksumMismatchError: SHA256 mismatch
Expected: 3b1634fa348c0f0064e43f3fb673e30a5e46f2b51cd6cd0d4a0cbf326e71c90e
  Actual: c150aa81c8fbbce15af3843ad0b8c048170faac6c933ef1a44bd40dc0cd1e2f0
 Archive: /Users/apple/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch
To retry an incomplete download, remove the file above.`

I am also attaching the Patch file if its gonna help it out.
![Patchfile](https://user-images.githubusercontent.com/54210493/63215770-1f18c080-c149-11e9-8983-67f45b024aba.png)
![Patchfile_contd](https://user-images.githubusercontent.com/54210493/63215771-1fb15700-c149-11e9-9c00-1f6cb6f11b3d.png)

**Comment by RgRaam** [comment]
I just got it. I

> > @RgRaam edit the patch file, should be `/Users/apple/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch` with nano, vim or whatever editor you'd like.
> > The only change you need to make is in the last line where it says "2.22.0" , change that to "2.21.0" (changing the 22 to 21)
> > Save and re-run your brew install
> 
> Thank You for the Reply @hnry . But it din't Solve. It came with another mismatch error.
> 
> ```
>  Downloading https://www.gnupg.org/ftp/gcrypt/gnutls/v3.6/gnutls-3.6.9.tar.xz
> Already downloaded: /Users/apple/Library/Caches/Homebrew/downloads/10375e457c78424d079bc03dc50b19e2d8f153be4087edde62a035a2627d8c6f--gnutls-3.6.9.tar.xz
> ==> Downloading https://gitlab.com/gnutls/gnutls/commit/ef80617d1e17e0878a909baa
> Already downloaded: /Users/apple/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch
> Error: An exception occurred within a child process:
>   ChecksumMismatchError: SHA256 mismatch
> Expected: 3b1634fa348c0f0064e43f3fb673e30a5e46f2b51cd6cd0d4a0cbf326e71c90e
>   Actual: c150aa81c8fbbce15af3843ad0b8c048170faac6c933ef1a44bd40dc0cd1e2f0
>  Archive: /Users/apple/Library/Caches/Homebrew/downloads/0c113ab0098bc4c3f2e8f410741b6104f2928fe63eb38cf98f8dcc3362f02ac4--ef80617d1e17e0878a909baad62a75ba265c0e00.patch
> To retry an incomplete download, remove the file
> ```
> 
> I am also attaching the Patch file if its gonna help it out.
> 
> ```
> From ef80617d1e17e0878a909baad62a75ba265c0e00 Mon Sep 17 00:00:00 2001
> From: Nikos Mavrogiannopoulos <nmav@gnutls.org>
> Date: Fri, 2 Aug 2019 21:57:40 +0200
> Subject: [PATCH] read_cpuid_vals: use __get_cpuid_count() only when available
> 
> This makes the functionality available on gcc 4.8.
> 
> Resolves: #812
> 
> Signed-off-by: Nikos Mavrogiannopoulos <nmav@gnutls.org>
> ---
>  configure.ac                     | 11 +++++++++++
>  lib/accelerated/x86/x86-common.c | 24 ++++++++++++++++++++----
>  2 files changed, 31 insertions(+), 4 deletions(-)
> 
> diff --git a/configure.ac b/configure.ac
> index 6a1d3dbc5c..1bf9bce95e 100644
> --- a/configure.ac
> +++ b/configure.ac
> @@ -182,6 +182,17 @@ case $host_cpu in
>    ;;
>  esac
> 
> +# check for gcc's __get_cpuid_count functionality
> +AC_MSG_CHECKING([for __get_cpuid_count])
> +AC_LINK_IFELSE(
> +   [AC_LANG_SOURCE([
> +    #include <cpuid.h>
> +    int main(void) { unsigned t1; return __get_cpuid_count(7, 0, &t1, &t1, &t1, &t1); }
> +   ])],
> +   [AC_DEFINE([HAVE_GET_CPUID_COUNT], [1], [use __get_cpuid_count]) AC_MSG_RESULT([yes])],
> +   [AC_MSG_RESULT([no])]
> +)
> +
>  fi
> 
>  AC_ARG_ENABLE(tls13-interop,
> diff --git a/lib/accelerated/x86/x86-common.c b/lib/accelerated/x86/x86-common.c
> index fb3ff90919..516d6776c5 100644
> --- a/lib/accelerated/x86/x86-common.c
> +++ b/lib/accelerated/x86/x86-common.c
> @@ -106,17 +106,33 @@ unsigned int _gnutls_x86_cpuid_s[4];
>  #define VIA_PADLOCK_PHE (1<<21)
>  #define VIA_PADLOCK_PHE_SHA512 (1<<22)
> 
> +#ifndef HAVE_GET_CPUID_COUNT
> +static inline void
> +get_cpuid_level7(unsigned int *eax, unsigned int *ebx,
> +		 unsigned int *ecx, unsigned int *edx)
> +{
> +	/* we avoid using __get_cpuid_count, because it is not available with gcc 4.8 */
> +	if (__get_cpuid_max(7, 0) < 7)
> +		return;
> +
> +	__cpuid_count(7, 0, *eax, *ebx, *ecx, *edx);
> +	return;
> +}
> +#else
> +# define get_cpuid_level7(a,b,c,d) __get_cpuid_count(7, 0, a, b, c, d)
> +#endif
> +
>  static unsigned read_cpuid_vals(unsigned int vals[4])
>  {
>  	unsigned t1, t2, t3;
> -	if (!__get_cpuid(1, &t1, &vals[0],
> -			 &vals[1], &t2))
> +	vals[0] = vals[1] = vals[2] = vals[3] = 0;
> +
> +	if (!__get_cpuid(1, &t1, &vals[0], &vals[1], &t2))
>  		return 0;
>  	/* suppress AVX512; it works conditionally on certain CPUs on the original code */
>  	vals[1] &= 0xfffff7ff;
> 
> -	if (!__get_cpuid_count(7, 0, &t1, &vals[2], &t2, &t3))
> -		return 0;
> +	get_cpuid_level7(&t1, &vals[2], &t2, &t3);
> 
>  	return 1;
>  }
> --
> 2.21.0
> ```
 I just Got it. I run `brew install https://raw.githubusercontent.com/smtcrms/linuxbrew-core/0cfd31347162db4fe5733d866a2475b8d79bb36f/Formula/gnutls.rb` . It seems my mac had a problem with the version of gnutls which was 3.6.9. Now i downloaded 3.6.8 and it works. ;) 

Damn! it took me 4days. ufffff

**Comment by kadnan** [comment]
> Thanks @jkkm. This worked like a charm.
> 
> Just remove the existing downloaded patch and copy this to the same file name and everybody goes home happy :)

@Varshul Can you please elaborate "the same file" part? Should I make an edit by doing `brew edit gnutls`?
